8325342: Remove unneeded exceptions in compare.sh

magicus opened the pull request:
Over time, we have been better at addressing inconsistencies in the build, but the exceptions put in place in compare.sh have not been updated to reflect this.

This attempts to make sure we only keep those exceptions that are currently actually needed to verify a reproducible build (the primary usecase for the compare script these days).

I also discovered bugs in the script, where mismatches in a zip file did not make us report a failed comparison. I added extra code to make sure we always set the REGRESSION flag whenever we detect a problem.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8325342](https://bugs.openjdk.org/browse/JDK-8325342): Remove unneeded exceptions in compare.sh (**Bug** - P4)


### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17735/head:pull/17735` \
`$ git checkout pull/17735`

Update a local copy of the PR: \
`$ git checkout pull/17735` \
`$ git pull https://git.openjdk.org/jdk.git pull/17735/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17735`

View PR using the GUI difftool: \
`$ git pr show -t 17735`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17735.diff">https://git.openjdk.org/jdk/pull/17735.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17735#issuecomment-1930364651)

erikj79 commented on make/scripts/compare.sh:
```diff
@@ -319,15 +292,10 @@ compare_general_files() {
-        ! -name "*.tar.gz" ! -name "gtestLauncher" \
-        ! -name "*.map" \
-        | $GREP -v "./bin/"  | $SORT | $FILTER)
+    GENERAL_FILES=$(cd $THIS_DIR && $FIND . -type f ! -name "*.pdb" \
+        | $GREP -v "/hotspot/gtest/server/gtestLauncher" \
+        | $GREP -v "/hotspot/gtest/server/libjvm.dylib" \
+        | $SORT | $FILTER)
```
This list of exclusion, or at least most of it, isn't for handling exceptions but because those files are handled explicitly by other comparison functions.

magicus replied:
I see. I'll split it into two parts then, one that deselects files that are guaranteed to be processed elsewhere, and one to filter out problematic files (currently only the gtest binaries and *.pdb).

magicus commented:
I will actually rework how files are selected completely.

magicus commented:
Ok, new try. I have redesigned how we discover files. Now we start by finding *all* files that we should compare, and then we partition out libs, zips, etc. What remains are the "other" files, so we will be sure to always check every file. I split up "other" files in two parts, debug symbol files and the "other" files proper.

I have also removed a few more exceptions I found. I have run this like 7-8 times in our CI system to check that we did not get any intermittent failures. They might still be possible, but at least they are somewhat rare. If they show up, we'll just have to reinstantiate the exception, and perhaps document it better, so we know when we can actually remove it.

magicus commented:
/integrate
